lib/src/nightly_shims.rs

```diff
@@ -0,0 +1,92 @@
+#[cfg(feature = "map_first_last")]
+pub trait BTreeMapExt<K, V> {
+    fn first_key(&self) -> Option<&K>;
+    fn last_key(&self) -> Option<&K>;
+    fn pop_first_value(&mut self) -> Option<V>;
+    fn pop_last_value(&mut self) -> Option<V>;
+}
+
+#[cfg(feature = "map_first_last")]
+impl<K: Ord + Clone, V> BTreeMapExt<K, V> for std::collections::BTreeMap<K, V> {
+    fn first_key(&self) -> Option<&K> {
+        self.keys().next()
+    }
+
+    fn last_key(&self) -> Option<&K> {
+        self.keys().next_back()
+    }
+    fn pop_first_value(&mut self) -> Option<V> {
+        self.first_entry().map(|x| x.remove())
+    }
+
+    fn pop_last_value(&mut self) -> Option<V> {
+        self.last_entry().map(|x| x.remove())
+    }
+}
+
+#[cfg(not(feature = "map_first_last"))]
+pub trait BTreeMapExt<K, V> {
+    fn first_key(&self) -> Option<&K>;
+    fn last_key(&self) -> Option<&K>;
+    fn pop_first_key(&mut self) -> Option<K>;
+    fn pop_last_key(&mut self) -> Option<K>;
+    fn pop_first_value(&mut self) -> Option<V>;
+    fn pop_last_value(&mut self) -> Option<V>;
+}
+
+#[cfg(not(feature = "map_first_last"))]
+impl<K: Ord + Clone, V> BTreeMapExt<K, V> for std::collections::BTreeMap<K, V> {
+    fn first_key(&self) -> Option<&K> {
+        self.keys().next()
+    }
+
+    fn last_key(&self) -> Option<&K> {
+        self.keys().next_back()
+    }
+
+    fn pop_first_key(&mut self) -> Option<K> {
+        let key = self.first_key()?;
+        let key = key.clone(); // ownership hack
```

Comment:
Note that this pattern is less performant, which is why I left it so that we could use nightly for the best implementation.